CBinding/Parser/CDocumentParser.cs

```diff
@@ -25,58 +25,111 @@
 // THE SOFTWARE.
 
 using System;
-using System.IO;
 using System.Collections.Generic;
-using System.Text.RegularExpressions;
-
-using MonoDevelop.Core;
-using MonoDevelop.Projects;
-using MonoDevelop.Ide;
 using MonoDevelop.Ide.TypeSystem;
-using ICSharpCode.NRefactory.TypeSystem;
-using ICSharpCode.NRefactory.TypeSystem.Implementation;
-using MonoDevelop.Core.Text;
+using ClangSharp;
+using MonoDevelop.Ide.Gui;
+using MonoDevelop.Ide.Editor;
+using System.Diagnostics;
+using System.Threading;
 
 namespace CBinding.Parser
 {
-	/// <summary>
-	/// Ctags-based document parser helper
-	/// </summary>
-	public class CDocumentParser:  TypeSystemParser
-	{
-		public override System.Threading.Tasks.Task<ParsedDocument> Parse (ParseOptions options, System.Threading.CancellationToken cancellationToken)
+	
+	public class CParsedDocument : DefaultParsedDocument {
+		public CXTranslationUnit TU { get; set;}
+		public CLangManager Manager { get; private set;}
+		public CProject Project { get; set;}
+		List<CXUnsavedFile> unsavedFiles;
+
+
+		public CParsedDocument(CProject proj, string fileName) : base(fileName)
 		{
-			var fileName = options.FileName;
-			var project = options.Project;
-			var doc = new DefaultParsedDocument (fileName);
-			doc.Flags |= ParsedDocumentFlags.NonSerializable;
-			ProjectInformation pi = ProjectInformationManager.Instance.Get (project);
-			
-			string content = options.Content.Text;
-			string[] contentLines = content.Split (new string[]{Environment.NewLine}, StringSplitOptions.None);
-			
-			var globals = new DefaultUnresolvedTypeDefinition ("", GettextCatalog.GetString ("(Global Scope)"));
-			lock (pi) {
-				// Add containers to type list
-				foreach (LanguageItem li in pi.Containers ()) {
-					if (null == li.Parent && FilePath.Equals (li.File, fileName)) {
-						var tmp = AddLanguageItem (pi, globals, li, contentLines) as IUnresolvedTypeDefinition;
-						if (null != tmp){ /*doc.TopLevelTypeDefinitions.Add (tmp);*/ }
+			Project = proj;
+			Manager = proj.cLangManager;
+			unsavedFiles = new List<CXUnsavedFile> ();
+			foreach (Document openDocument in MonoDevelop.Ide.IdeApp.Workbench.Documents) {
+				if (openDocument.IsDirty) {
+					CXUnsavedFile unsavedFile = new CXUnsavedFile ();
+					unsavedFile.Filename = openDocument.FileName;
+					unsavedFile.Length = openDocument.Editor.Text.Length;
+					unsavedFile.Contents = openDocument.Editor.Text;
+					if (Project.BOMPresentInFile [openDocument.FileName]) {
```

Comment:
This is the second place I've seen this exact code, and it's not immediately obvious why it's necessary. I would factor it out to a helper method with a descriptive name and comment.